[torchbind] add tracing_mode support

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #129586


Sometimes, it could be difficult to write a fake class e.g. when the original implementation is using some third-party libraries or users are certain that the class is safe to trace with the real object. 

This PR allows user to specify their intention by implementing a "safe_to_trace_with_real_obj" method on their script class.

Test Plan:
`pytest test/export/test_torchbind.py -k safe`


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

ydwu4 committed:
[torchbind] add safe_to_trace_with_real_obj support

[ghstack-poisoned]

ydwu4 committed:
Update on "[torchbind] add safe_to_trace_with_real_obj support"



Sometimes, it could be difficult to write a fake class e.g. when the original implementation is using some third-party libraries or users are certain that the class is safe to trace with the real object. 

This PR allows user to specify their intention by implementing a "safe_to_trace_with_real_obj" method on their script class.

Test Plan:
`pytest test/export/test_torchbind.py -k safe`


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

ydwu4 commented on torch/_dynamo/output_graph.py:
```diff
@@ -1577,14 +1577,19 @@ def update_used_symbols(used_symbols, fake: Union[torch.SymInt, torch.Tensor]):
-                                lambda t: update_used_symbols(used_symbols, t),
-                                fake_attr_val,
-                            )
+                        if not torch._library.fake_class_registry.tracing_with_real(
```
Tracing with real will always gives static shape and there is no way we could peek into its contents and get all the symbols. So we can just skip.

ydwu4 committed:
Update on "[torchbind] add safe_to_trace_with_real_obj support"



Sometimes, it could be difficult to write a fake class e.g. when the original implementation is using some third-party libraries or users are certain that the class is safe to trace with the real object. 

This PR allows user to specify their intention by implementing a "safe_to_trace_with_real_obj" method on their script class.

Test Plan:
`pytest test/export/test_torchbind.py -k safe`


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

zou3519 commented on torch/_library/fake_class_registry.py:
```diff
@@ -99,9 +99,18 @@ def _check_valid_flat_script_obj(flat_x):
 
-def to_fake_obj(fake_mode, x: torch.ScriptObject) -> FakeScriptObject:
+def tracing_with_real(x: torch.ScriptObject) -> bool:
+    return hasattr(x, "safe_to_trace_with_real_obj") and x.safe_to_trace_with_real_obj()
```
API bikeshedding:

What if we just had a "tracing_mode" flag that was either "real" or "fake"?

zou3519 commented on torch/_higher_order_ops/torchbind.py:
```diff
@@ -106,12 +106,20 @@ def inner(mode, *args, **kwargs):
 
-# TODO: currently we just run the C++ implementation with fake tensors.
-# But we should make it possible to register a fake torchbind implementation.
+# When tracing with fake script object, the call_torchbind op will return a fake tensor
+# When tracing with real script object, the call_torchbind op return a real tensor,
+# we need to convert it to real tensor mannually.
```
What's the strategy for tracing with a real script object?
- We run methods with the real script object and FakeTensors? The alternative is running methods with the real script object and real Tensors.
- We run operators with the real script objects and FakeTensors? The alternative is running operators with real script objects and real Tensors

ydwu4 replied:
> we run methods and operators with real script object and fake tensor inputs 

This is the current status. People implemented meta kernels that handles the real script object + fake tensor inputs.

> The alternative is running methods with the real script object and real Tensors.

Is it possible to get a real tensors while we're tracing a torchbind method/operator? We need to trace it together with real tensor inputs along with fake tensor inputs, right?

zou3519 replied:
Thanks for clarifying, can you put down the current approach somewhere?

ydwu4 replied:
Sure!

ydwu4 committed:
Update on "[torchbind] add safe_to_trace_with_real_obj support"



Sometimes, it could be difficult to write a fake class e.g. when the original implementation is using some third-party libraries or users are certain that the class is safe to trace with the real object. 

This PR allows user to specify their intention by implementing a "safe_to_trace_with_real_obj" method on their script class.

Test Plan:
`pytest test/export/test_torchbind.py -k safe`


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

ydwu4 committed:
Update on "[torchbind] add tracing_mode support"



Sometimes, it could be difficult to write a fake class e.g. when the original implementation is using some third-party libraries or users are certain that the class is safe to trace with the real object. 

This PR allows user to specify their intention by implementing a "safe_to_trace_with_real_obj" method on their script class.

Test Plan:
`pytest test/export/test_torchbind.py -k safe`


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

zou3519 commented on test/cpp/jit/test_custom_class_registrations.cpp:
```diff
@@ -222,6 +222,21 @@ struct TensorQueue : torch::CustomClassHolder {
+    return x_;
+  }
+
+  std::string tracing_mode() {
```
Need to rename your commit message  and the PR body to say "tracing_mode"

zou3519 commented on torch/_library/fake_class_registry.py:
```diff
@@ -99,10 +99,26 @@ def _check_valid_flat_script_obj(flat_x):
+    return x.tracing_mode() == "real"
+
+
+def to_fake_obj(
```
if it sometimes converts to fake obj, then we should name this maybe_to_fake_obj or "to_traceable_obj"

zou3519 commented on torch/_library/fake_class_registry.py:
```diff
@@ -99,10 +99,26 @@ def _check_valid_flat_script_obj(flat_x):
 
-def to_fake_obj(fake_mode, x: torch.ScriptObject) -> FakeScriptObject:
+def tracing_with_real(x: torch.ScriptObject) -> bool:
+    if not hasattr(x, "tracing_mode"):
```
Do you think people want all instances of an e.g. TensorQueue to trace with real, or they might want to do different behavior for specific TensorQueue?

If it's all then tracing_mode should be a classmethod. This might be the simplest for now.

ydwu4 committed:
Update on "[torchbind] add tracing_mode support"



Sometimes, it could be difficult to write a fake class e.g. when the original implementation is using some third-party libraries or users are certain that the class is safe to trace with the real object. 

This PR allows user to specify their intention by implementing a "safe_to_trace_with_real_obj" method on their script class.

Test Plan:
`pytest test/export/test_torchbind.py -k safe`


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

ydwu4 commented:
@pytorchbot merge

ydwu4 commented:
@pytorchbot merge
